### PR TITLE
Exponent f32

### PR DESCRIPTION
**Pull request by 4ekmah**
Fixing all register allocator problems caused by introducing vector SIMD registers.
Exponent implementation and test.
Prohibition of complex conditions is now verbosed.

### CODE REVIEW

File: include/loops/loops.hpp [hunk 1]
```diff
@@ -33,6 +33,7 @@ enum {
     OP_STORE,
 
     OP_MOV,
+    OP_MOVK,   //Move bytes to shifted byte position of register and keep other bits unchanged.
```
Comment (vpisarev) [inline include/loops/loops.hpp:36]: it's ARM-specific operation, right? Can you add a proper prefix to it? OP_AARCH64_MOVK or something like that

File: test/test_math.hpp [hunk 2]
```diff
@@ -86,11 +87,43 @@ PTESTfix_2(exponentiation_by_squaring_v, uint32_t, 9);
...
+    USE_CONTEXT_(CTX);
+    STARTFUNC_(TESTNAME, &dest, &src, &n)
+    {
+        auto expc = expInitConsts(CTX);
+        IReg offset = CONST_(0);
+        WHILE_(n > 0)
+        {
+            VReg<float> x = loadvx<float>(src,offset);
+            storevx(dest, offset, exp(x, expc));
+            offset += CTX.vectorRegisterSize();
```
Comment (vpisarev) [inline test/test_math.hpp:108]: Need to shorter name for 'vectorRegisterSize()'

File: test/test_math.hpp [hunk 3]
```diff
@@ -86,11 +87,43 @@ PTESTfix_2(exponentiation_by_squaring_v, uint32_t, 9);
...
+LTEST(exp_f32, {
+    IReg dest, src, n;
+    USE_CONTEXT_(CTX);
+    STARTFUNC_(TESTNAME, &dest, &src, &n)
+    {
+        auto expc = expInitConsts(CTX);
+        IReg offset = CONST_(0);
+        WHILE_(n > 0)
+        {
+            VReg<float> x = loadvx<float>(src,offset);
```
Comment (vpisarev) [inline test/test_math.hpp:106]: maybe loadvx -> loadvec?

File: test/test_math.hpp [hunk 4]
```diff
@@ -86,11 +87,43 @@ PTESTfix_2(exponentiation_by_squaring_v, uint32_t, 9);
...
+PTESTfix_2(exponentiation_by_squaring_v, double, 0);
 PTESTfix_2(exponentiation_by_squaring_v, double, 4);
 PTESTfix_2(exponentiation_by_squaring_v, double, 9);
+
+LTEST(exp_f32, {
+    IReg dest, src, n;
+    USE_CONTEXT_(CTX);
+    STARTFUNC_(TESTNAME, &dest, &src, &n)
+    {
+        auto expc = expInitConsts(CTX);
```
Comment (vpisarev) [inline test/test_math.hpp:102]: expInitConsts -> expInit()?

File: src/common.cpp [hunk 5]
```diff
@@ -365,6 +372,67 @@ namespace loops
 
     __Loops_FuncScopeBracket_::~__Loops_FuncScopeBracket_() { CTX->endFunc(); }
 
+    exp_consts::exp_consts(VReg<float>&& a_lo, VReg<float>&& a_hi, VReg<float>&& a_half, VReg<float>&& a_one,
```
Comment (vpisarev) [inline src/common.cpp:375]: do we need such a constructor?

File: include/loops/loops.hpp [hunk 6]
```diff
@@ -783,5 +837,23 @@ template<typename _Tp> VReg<_Tp> pow(const VReg<_Tp>& a, int p)
...
 }
 
+struct exp_consts
+{
+    VReg<float> lo, hi, half, one, LOG2EF, C1, C2, p0, p1, p2, p3, p4, p5;
+    VReg<int32_t> _7f;
+    exp_consts(VReg<float>&& a_lo, VReg<float>&& a_hi, VReg<float>&& a_half, VReg<float>&& a_one,
+                        VReg<float>&& a_LOG2EF, VReg<float>&& a_C1, VReg<float>&& a_C2, VReg<float>&& a_p0,
+                        VReg<float>&& a_p1, VReg<float>&& a_p2, VReg<float>&& a_p3, VReg<float>&& a_p4,
+                        VReg<float>&& a_p5, VReg<int32_t>&& a_7f);
```
Comment (vpisarev) [inline include/loops/loops.hpp:847]: maybe this constructor is not needed. There should also be default constructor, perhaps

File: include/loops/loops.hpp [hunk 7]
```diff
@@ -676,21 +701,48 @@ template<typename _Tp> VReg<_Tp>& operator ^= (VReg<_Tp>& a, const VReg<_Tp>& b)
...
+    return newiopV<_Dp>(VOP_TRUNC, {a});
+}
+template<typename _Dp> VReg<_Dp> trunc(const VReg<float>& a)  //Convert with rounding to zero
+{
+    static_assert(sizeof(_Dp) == sizeof(float), "Attempt to convert real number to integer of different size.");
+    return newiopV<_Dp>(VOP_TRUNC, {a});
+}
+template<typename _Dp> VReg<_Dp> trunc(const VReg<double>& a)
+{
+    static_assert(sizeof(_Dp) == sizeof(double), "Attempt to convert real number to integer of different size.");
```
Comment (vpisarev) [inline include/loops/loops.hpp:717]: can we just use VReg<int64_t>?

File: include/loops/loops.hpp [hunk 8]
```diff
@@ -676,21 +701,48 @@ template<typename _Tp> VReg<_Tp>& operator ^= (VReg<_Tp>& a, const VReg<_Tp>& b)
...
-    VReg<_Dp> res;
-    res.func = a.func;
-    res.idx = a.idx;
-    return res;
+    static_assert(sizeof(_Dp) == sizeof(f16_t), "Attempt to convert real number to integer of different size.");
+    return newiopV<_Dp>(VOP_TRUNC, {a});
+}
+template<typename _Dp> VReg<_Dp> trunc(const VReg<float>& a)  //Convert with rounding to zero
+{
+    static_assert(sizeof(_Dp) == sizeof(float), "Attempt to convert real number to integer of different size.");
```
Comment (vpisarev) [inline include/loops/loops.hpp:712]: can we just use VReg<int32_t>?

**Comment by vpisarev** [comment]
Can you also add a test where complex condition is used? The test should include try-catch and produce a failure if there is no exception thrown.

File: include/loops/loops.hpp (hunk 8 above)
Comment (4ekmah) [reply include/loops/loops.hpp:712]: It also supports float -> uint32_t

File: include/loops/loops.hpp (hunk 7 above)
Comment (4ekmah) [reply include/loops/loops.hpp:717]: It also supports double -> uint64_t

File: include/loops/loops.hpp (hunk 1 above)
Comment (4ekmah) [reply include/loops/loops.hpp]: Done.

File: test/test_math.hpp (hunk 2 above)
Comment (4ekmah) [reply test/test_math.hpp]: Done.

File: test/test_math.hpp (hunk 3 above)
Comment (4ekmah) [reply test/test_math.hpp]: Done.

File: test/test_math.hpp (hunk 4 above)
Comment (4ekmah) [reply test/test_math.hpp]: Done.

File: include/loops/loops.hpp (hunk 6 above)
Comment (4ekmah) [reply include/loops/loops.hpp]: There cannot be default constructor, it's needed at least Context.
That's why this constructor was replaced with exp_consts(Context CTX);

**Pull request merged**